### DISCUSSION TITLE
docker behind nginx , websockify forwarding to docker to enable sound

### DISCUSSION THREAD

**Discussion by Akatsuki-Ryu** [jlesage/docker-firefox#262] [Q&A] [answered]
hello , thank you very much for this docker image. i deployed it on my personal home server but I need to work with an nginx to route this docker container under a certain path . 

i referred to your instructions about the situation what the container behind nginx. and I applied this code (similar) to my nginx settings 

```
map $http_upgrade $connection_upgrade {
	default upgrade;
	''      close;
}

upstream docker-firefox {
	# If the reverse proxy server is not running on the same machine as the
	# Docker container, use the IP of the Docker host here.
	# Make sure to adjust the port according to how port 5800 of the
	# container has been mapped on the host.
	server 127.0.0.1:5800;
}

server {
	[...]

	location = /firefox {return 301 $scheme://$http_host/firefox/;}
	location /firefox/ {
		proxy_pass http://docker-firefox/;
		# Uncomment the following line if your Nginx server runs on a port that
		# differs from the one seen by external clients.
		#port_in_redirect off;
		location /firefox/websockify {
			proxy_pass http://docker-firefox/websockify/;
			proxy_http_version 1.1;
			proxy_set_header Upgrade $http_upgrade;
			proxy_set_header Connection $connection_upgrade;
			proxy_read_timeout 86400;
		}
	}
}
```

the docker container started and I can see the firebox UI from localhost:port/firefox/ , but there is no sound . when I checked the terminal log , it is calling localhost:port/websockify-audio instead of going under /firefox/websockify as I might expect.  which resulted in so sound in the browser. 

i could have done the following code in my nginx settings to also forward the websockify to that container.
```
        location /websockify {
            proxy_pass http://docker:5800; 
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
            proxy_read_timeout 86400;
        }
```
but my problem is that the firefox docker container's port is different everytime as I want to setup multiple firefox docker containers at the same time using different port. 

so I am wondering if there is any suggestions or advises , any idea is welcome . much appreciated !

**Comment by toomyzoom** [comment, marked as answer]
So the main VNC server (noVNC) uses websocket library (websock.js) to connect. It uses `this._url`, which is whatever your current address, e.g.: `wss://<domain>:443/firefox/websockify`
The audio call for connecting is in noVNC (app/ui.js), in this function: `connectWebSocket()`
```
function connectWebSocket() {
  var socketURL = (window.location.protocol == 'https:' ? 'wss:' : 'ws:')+'//'+document.location.hostname+':'+window.location.port+'/websockify-audio';
  ...
  UI.audioContext.webSocket = new WebSocket(socketURL, ['binary']);
```

It hardcoded the url as `document.location.hostname+':'+window.location.port+'/websockify-audio'` so we get `wss://<domain>:443/websockify-audio` instead of `wss://<domain>:443/firefox/websockify-audio`

**Reply by Akatsuki-Ryu** [reply]
thank you very much for the answer . i also noticed that part of the code as I am digging deeper . it would be nice to have a ENV to indicate what is the path of the socketURL for the reverse proxy . lets see if I have time to make a pull request

**Comment by jlesage** [comment]
Thank you for reporting. This is a bug that will be fixed in the baseimage.

**Reply by Akatsuki-Ryu** [reply]
that would be fantastic ! thank you very much
